Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -93,9 +119,92 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
         *cost_real = 0.;
     }
 #endif
-    amrex::ParallelFor(
-        np_to_depose,
-        [=] AMREX_GPU_DEVICE (long ip) {
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    const auto dxiarr = geom.InvCellSizeArray();
+    const auto plo = geom.ProbLoArray();
+    const auto domain = geom.Domain();
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
+
+    const auto npts = sample_tbox_x.numPts();
+
+    const int nblocks = a_bins.numBins();
+    const auto offsets_ptr = a_bins.offsetsPtr();
+    const int threads_per_block = 256;
+
+    std::size_t shared_mem_bytes = npts*sizeof(amrex::Real);
+    amrex::IntVect bin_size = WarpX::sort_bin_size;
+
+    const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
+
+    if (shared_mem_bytes > max_shared_mem_bytes) {
+      std::cout << "Tile size for charge deposition of " << shared_mem_bytes
+                << " bytes would not fit in GPU's shared memory of " << max_shared_mem_bytes << " bytes" << std::endl;
```

Comment:
Let's use `amrex::Print()` or even better our warning logger here:
https://warpx.readthedocs.io/en/latest/developers/warning_logger.html

Reply:
Discussed with @atmyers: since the shared memory version also requires sorting, we might want to add an additional runtime option to  control this also explicitly.